src/steamship/agents/schema/context.py

```diff
@@ -54,6 +58,132 @@ def get_or_create(
+            ctx_file = File.create(client=client, content="agent context file", handle=handle)
+
+        for block in ctx_file.blocks:
+            for tag in block.tags:
+                if tag.kind == "AgentContext":
+                    block.delete()
+
+        Block.create(
+            client=client,
+            file_id=ctx_file.id,
+            text=json.dumps(self.metadata),
+            tags=[Tag(kind="AgentContext", name="metadata")],
+        )
+
+        chat_history_tags = self.chat_history.tags
+        for tag in chat_history_tags:
+            if tag.kind == TagKind.CHAT and tag.name == ChatTag.CONTEXT_KEYS:
+                ctx_keys = tag.value
+
+        Block.create(
+            client=client,
+            file_id=ctx_file.id,
+            text=json.dumps(ctx_keys),
+            tags=[Tag(kind="AgentContext", name="context_keys")],
+        )
+
+        steps = self.completed_steps
+        # TODO(dougreid): this is lazy. really, we need both UUID and text for the text case.
+        json_steps = []
+        for step in steps:
+            json_inputs = []
+            for block in step.input:
+                if block.is_text():
+                    json_inputs.append({"text": block.text})
+                else:
+                    json_inputs.append({"uuid": block.id})
+
+            json_outputs = []
+            for block in step.output:
+                if block.is_text():
+                    json_outputs.append({"text": block.text})
+                else:
+                    json_outputs.append({"uuid": block.id})
+
+            json_step = {
+                "tool": step.tool.name,
+                "input": json_inputs,
+                "output": json_outputs,
+            }
+            json_steps.append(json_step)
+
+        Block.create(
+            client=client,
+            file_id=ctx_file.id,
+            text=json.dumps(json_steps),
+            tags=[Tag(kind="AgentContext", name="completed_steps")],
+        )
+
+        return handle
+
+    @staticmethod
+    def hydrate_from_file(client: Steamship, file_handle: str):
+        ctx_file = File.get(client=client, handle=file_handle)
+        if not ctx_file:
+            raise SteamshipError("context not found")
+
+        new_ctx = AgentContext()
+        new_ctx.completed_steps = []
+        new_ctx.client = client
+
+        for block in ctx_file.blocks:
+            for tag in block.tags:
+                if tag.kind == "AgentContext":
+                    if tag.name == "metadata":
+                        new_ctx.metadata = json.loads(block.text)
+                    elif tag.name == "context_keys":
+                        from steamship.agents.schema.chathistory import ChatHistory
+
+                        new_ctx.chat_history = ChatHistory.get_or_create(
+                            client=client, context_keys=json.loads(block.text)
+                        )
+                    elif tag.name == "completed_steps":
```

Comment:
I think I need to do more learning on diff between the steps & the chat history --- I have a blurry understanding of when something wants to be in one and when in the other